analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs

```diff
@@ -0,0 +1,133 @@
+        private const string MessageFormat = "Make sure publicly writable directories are used safely here.";
+        private readonly DiagnosticDescriptor rule;
+
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+
+        private protected abstract bool IsGetTempPathAssignment(TInvocationExpression invocationExpression, KnownType type, string methodName, SemanticModel semanticModel);
+        private protected abstract bool IsInsecureEnvironmentVariableRetrieval(TInvocationExpression invocation, KnownType type, string methodName, SemanticModel semanticModel);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected string[] InsecureEnvironmentVariables { get; } = { "tmp", "temp", "tmpdir" };
+
+        protected PubliclyWritableDirectoriesBase(IAnalyzerConfiguration configuration, ResourceManager rspecResources) : base(configuration) =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources).WithNotConfigurable();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    if (!IsEnabled(c.Options))
+                    {
+                        return;
+                    }
+
+                    var node = c.Node;
+                    if (Language.Syntax.NodeStringTextValue(node) is { } stringValue
+                        && PubliclyWritableDirectoriesRegexMatcher.IsSensitiveDirectoryUsage(stringValue))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, node.GetLocation()));
+                    }
+                },
+                Language.SyntaxKind.StringLiteralExpression,
+                Language.SyntaxKind.InterpolatedStringExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    if (!IsEnabled(c.Options))
+                    {
+                        return;
+                    }
+
+                    var node = c.Node;
+                    if (node is TInvocationExpression invocation
+                        && (IsGetTempPathAssignment(invocation, KnownType.System_IO_Path, "GetTempPath", c.SemanticModel)
+                            || IsInsecureEnvironmentVariableRetrieval(invocation, KnownType.System_Environment, "GetEnvironmentVariable", c.SemanticModel)))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, node.GetLocation()));
+                    }
+                },
+                Language.SyntaxKind.InvocationExpression);
+        }
+
+        private static class PubliclyWritableDirectoriesRegexMatcher
```

Comment:
As we're inside `PubliclyWritableDirectoriesBase` scope
```suggestion
        private static class RegexMatcher
```